src/vos/vos_internal.h

```diff
@@ -1029,4 +1029,111 @@ oi_iter_aggregate(daos_handle_t ih, bool discard);
 int
 vos_obj_iter_aggregate(daos_handle_t ih, bool discard);
 
+
+/* EV tree physical entry */
+struct vos_agg_phy_ent {
+	d_list_t		pe_link;
+	/* Original in-tree rectangle */
+	struct evt_rect		pe_rect;
+	/* Entry payload address */
+	bio_addr_t		pe_addr;
+	/* Original csums for entry */
+	struct dcs_csum_info	pe_csum_info;
+	/*
+	 * Extent start offset for the truncated physical entry. (Entry
+	 * straddles window end could be truncated on merge window flush)
+	 */
+	daos_off_t		pe_off;
+	/* Pool map version of entry */
+	uint32_t		pe_ver;
+	/* Referenced by visible entries */
+	uint32_t		pe_ref;
+	/* Need to truncate on window flush */
+	bool			pe_trunc_head;
+};
+
+/* EV tree logical entry */
+struct vos_agg_lgc_ent {
+	struct evt_extent	 le_ext;
+	struct vos_agg_phy_ent	*le_phy_ent;
+};
+
+/*
+ * EV tree logical segment, it'll be used to form new physical
+ * rectangle and being inserted in evtree on merge window flush.
+ */
+struct vos_agg_lgc_seg {
+	/* Start index in mw_lgc_ents */
+	unsigned int		 ls_idx_start;
+	/* End index in mw_lgc_ents */
+	unsigned int		 ls_idx_end;
+	/* New segment generated by truncate on window flush */
+	struct vos_agg_phy_ent	*ls_phy_ent;
+	/* Description of the new physical entry to be inserted */
+	struct evt_entry_in	 ls_ent_in;
+	/* indicator of csum error on verify */
+	bool			 ls_has_csum_err;
+};
+
+/* I/O context used on EV tree merge window flush */
+struct vos_agg_io_context {
+	/* Temporary buffer for data transfer */
+	void			*ic_buf;
+	unsigned int		 ic_buf_len;
+	/* Buffer to hold output csums for entire aggregation */
+	void			*ic_csum_buf;
+	unsigned int		 ic_csum_buf_len;
+	/* Segments being involved on merge window flush */
+	struct vos_agg_lgc_seg	*ic_segs;
+	unsigned int		 ic_seg_max;
+	unsigned int		 ic_seg_cnt;
+	/* Reserved SCM extents for new physical entries */
+	struct pobj_action	*ic_scm_exts;
+	unsigned int		 ic_scm_max;
+	unsigned int		 ic_scm_cnt;
+	/* Reserved NVMe extents for new physical entries */
+	d_list_t		 ic_nvme_exts;
+};
+
```

Comment:
Please keep these data structures in vos_aggregate.c, I don't think we should expose all these aggregation details. Some simpler neutral structures should be shared by vos_aggregate.c and vos_csum.c instead.